Paper: Fast Exploration of the Milky Way (or any other n-dimensional dataset) (#815)
scipy-conference/scipy_proceedings: papers/Francesc_Alted/Blosc2_NDim.rst

       return out

Finally, we can compute the density of stars in a 3D grid with this script:

.. code-block:: python

   R = 2  # resolution of the 3D cells in ly
   LY_RADIUS = 10_000  # radius of the sphere in ly
   CUBE_SIDE = (2 * LY_RADIUS) // R
   MAX_STARS = 1000_000_000  # max number of stars to load

   b = blosc2.open("gaia-ly.b2nd")
   x = b[0, :MAX_STARS]
   y = b[1, :MAX_STARS]
   z = b[2, :MAX_STARS]
   g = b[3, :MAX_STARS]

   # Create 3d array.
   # Be sure to have enough swap memory (around 4 TB!)
   a3d = np.zeros((CUBE_SIDE, CUBE_SIDE, CUBE_SIDE),
                  dtype=np.float32)
   for i, coords in enumerate(zip(x, y, z)):
       x_, y_, z_ = coords
       a3d[(int(x_) + LY_RADIUS) // R,
           (int(y_) + LY_RADIUS) // R,
           (int(z_) + LY_RADIUS) // R] += g[i]

   # Save 3d array as Blosc2 NDim file
   blosc2.asarray(a3d,
                  urlpath="gaia-3d.b2nd", mode="w",
                  chunks=(200, 200, 200),
                  blocks=(20, 20, 20),
                  )

With that, we have a 3D array of shape (10_000, 10_000, 10_000) with the magnitudes of stars with a 2 light years resolution.  We can visualize it with the following code:

> To be completed ...

Reviewer: Hi @FrancescAlted, do you have an estimate for when this will be completed?

Author: The visualization code is work in progress right now.  We will be working on it for the next 2 or 3 weeks or so.  Is that acceptable?

Reviewer: Thanks. As long as reviewers have time to complete their comments it is fine. Here are the reviewer deadlines: https://github.com/scipy-conference/scipy_proceedings#reviewer-deadlines.

:author: Project Blosc
:email:
:institution: Project Blosc
:equal-contributor:
:bibliography: mybib

:author: Francesc Alted
:email: francesc@blosc.org
:institution: Project Blosc
:equal-contributor:
:corresponding:

:author: Marta Iborra
:email: martaiborra24@gmail.com
:institution: Project Blosc
:equal-contributor:

:author: Oscar Guiñón
:email: soscargm98@gmail.com
:institution: Project Blosc
:equal-contributor:

:author: David Ibáñez
:email: jdavid.ibp@gmail.com
:institution: Project Blosc

:author: Sergio Barrachina
:email: barrachi@uji.es
:institution: Universitat Jaume I

---------------------------------------------------------------------------------

> Using Blosc2 NDim As A Fast Explorer Of The Milky Way (Or Any Other NDim Dataset)

Reviewer 2: The title draws attention to the exploration of the milky way rather focus on the algorithm. I would suggest something like "Blosc2 NDim: A Blosc extension to assist the efficient exploration of the high dimensional datasets"

---------------------------------------------------------------------------------

.. class:: abstract

>     N-dimensional datasets are widely used in many scientific fields. Quickly accessing subsets of these large datasets is critical for an efficient exploration experience. To achieve this goal, we have added support for multidimensional datasets to Blosc2, a compression and format library. This extended library can effectively deal with n-dimensional large sparse datasets, as the zeroed parts are almost entirely suppressed, while the non-zero parts are stored in smaller sizes than their uncompressed counterparts. In addition to this, the new two level data partition used in Blosc2 reduces the need for decompressing unnecessary data, which allows top-class slicing speed.

Reviewer 3: Here in the "N-dimensional dataset," N is not defined (N ca"N-dimensional dataset," N is not defined (N can be 2 or 3, which is not large). Instead, use the term "high-dimensional" everywhere wherever "n-dimensional" is used.

Reviewer 3: My Suggestion is to rewrite the abstract. I have provided an example. I would suggest the authors to rewrite in a similar way.
Abstract:
Large dimensional datasets have been widely used in various engineering and scientific applications. Prompt access to the subsets of these datasets is crucial for an efficient exploration experience. To facilitate this experience, we have added support for large dimensional datasets to Blosc2, a compression and format library. The extension to the library can effectively support large dimensional sparse datasets by suppressing the zero elements and compressing the non-zero elements. In addition to this, the new two-level data partition used in Blosc2 reduces the need for decompressing unnecessary data, further accelerating slicing speed.

Reviewer 3: I do not understand the meaning of "suppress" used here. Does it mean compress or deletion?

Author: > Here in the "N-dimensional dataset," N is not defined (N ca"N-dimensional dataset," N is not defined (N can be 2 or 3, which is not large). Instead, use the term "high-dimensional" everywhere wherever "n-dimensional" is used.

I think this depends on the area of use.  For example, high-dimensional in data science means something along these lines: [High-dimensional data refers to a dataset where the number of features f (or covariates) can be even larger than the number of observations N](https://www.sciencedirect.com/topics/computer-science/high-dimensional-data#:~:text=High%2Ddimensional%20data%20refers%20to,of%20observations%20N%20%5B105%5D.), which is not what we want to express.

Author: > My Suggestion is to rewrite the abstract. I have provided an example. I would suggest the authors to rewrite in a similar way.
> Abstract:
> Large dimensional datasets have been widely used in various engineering and scientific applications. Prompt access to the subsets of these datasets is crucial for an efficient exploration experience. To facilitate this experience, we have added support for large dimensional datasets to Blosc2, a compression and format library. The extension to the library can effectively support large dimensional sparse datasets by suppressing the zero elements and compressing the non-zero elements. In addition to this, the new two-level data partition used in Blosc2 reduces the need for decompressing unnecessary data, further accelerating slicing speed.

Ok.  I like your wording.  I used that in 9751365d

>     The Blosc2 NDim layer enables the creation and reading of n-dimensional datasets in an extremely efficient manner. This is due to a completely general n-dim 2-level partitioning, which allows for slicing and dicing of arbitrary large (and compressed) data in a more fine-grained way. Having a second partition provides a better flexibility to fit the different partitions at the different CPU cache levels, making compression even more efficient.

Reviewer 3: Again replace "n-dimensional" with "high dimensional"

    As an example, we will demonstrate how Blosc2 NDim enables fast exploration of the Milky Way using the Gaia DR3 dataset. This catalog contains information on 1.7 billion stars in our galaxy, but we have chosen to include just the stars that are in a sphere of 10 thousand light-years radius (centered in the Gaia telescope), which accounts for 0.7 billion stars. The total size of the dataset of star positions is 7.3 TB, but when compressed, it is reduced to just 6.5 GB, making it easy to fit into the memory of modern computers for being processed.

.. class:: keywords

    explore datasets, n-dimensional datasets, Gaia DR3, Milky Way, Blosc2, compression

Introduction
------------

> The exploration of n-dimensional datasets is a common practice in many areas of science. However, one of its drawbacks is that the explored datasets size can become very large, which will slow down the exploration process significantly. In this paper, we demonstrate how Blosc2 NDim can be used to accelerate the exploration of huge n-dimensional datasets.

Reviewer 3: Instead of "The exploration of n-dimensional datasets is a common practice in many areas of science. However, one of its drawbacks is that the explored datasets size can become very large, which will slow down the exploration process significantly" 

I will suggest the following: 
"The exploration of datasets that are high dimensional is a common practice in various fields of science. However, exploring such high-dimensional datasets is challenging when the memory size of the dataset is extremely large. This can slow down the data exploration process significantly"

Author: Ok.  Done in 3851456a

Blosc is a high-performance compressor optimized for binary data. Its design enables faster transmission of data to the processor cache than the traditional, non-compressed, direct memory fetch approach using an OS call to ``memcpy()``. This can be helpful not only in reducing the size of large datasets on-disk and in-memory, but also in accelerating memory-bound computations, which are typical in big data processing.

> Blosc uses the blocking technique :cite:`FA10-starving-cpus` to minimize activity on the memory bus. The technique divides datasets into blocks small enough to fit in the caches of modern processors, where compression/decompression is performed. Blosc also takes advantage of SIMD (SSE2, AVX2, NEON…) and multi-threading capabilities in modern multi-core processors to maximize the compression/decompression speed.

Reviewer 3: Mention the full form of SIMD, since used for the first time " single-instruction multiple-data streams (SIMD)"

Author: Fixed in 4d22d171

In addition, using the Blosc compressed data can accelerate memory-bound computations when enough cores are dedicated to the task. Figure :ref:`sum-precip` provides a real example of this.

.. figure:: sum_openmp-rainfall.png
   :scale: 40%

   Speed for summing up a vector of real float32 data (meteorological precipitation) using a variety of codecs provided by Blosc2. Note that the maximum speed is achieved when utilizing the maximum number of (logical) threads available on the computer (28), where different codecs are allowing faster computation than using uncompressed data. Benchmark performed on a Intel i9-10940X CPU, with 14 physical cores. More info at :cite:`BDT18-breaking-memory-walls`. :label:`sum-precip`

Blosc2 is the latest version of the Blosc 1.x series, which is used in many important libraries, such as HDF5 :cite:`hdf5`, Zarr :cite:`zarr`, and PyTables :cite:`pytables`. Its NDim feature excels at reading multi-dimensional slices, thanks to an innovative pineapple-style partitioning technique :cite:`BDT23-blosc2-ndim-intro`. This enables fast exploration of general n-dimensional datasets, including the 3D Gaia dataset.

> The Gaia dataset

Reviewer 3: I would suggest the name to be "Milky Way dataset" as the title. Later Gaia dataset can be mentioned

Author: Fixed in 4d5973ed

----------------

The Gaia DR3 dataset is a catalog containing information on 1.7 billion stars in our galaxy. For this work, we extracted the 3D coordinates of 1.4 billion stars (those with non-null parallax values). When stored as a binary table, the dataset is 22 GB in size (uncompressed).

> However, we converted the tabular dataset into a sphere with a radius of 10,000 light years and framed it into a 3D array of shape (20,000, 20,000, 20,000). Each cell in the array represents a cube of 1 light year per side and contains the number of stars within it. Given that the average distance between stars in the Milky Way is about 5 light years, very few cells will contain more than one star (e.g. the maximum of stars in a single cell in our sphere is 6). This 3D array contains 0.7 billion stars, which is a significant portion of the Gaia catalog.

Reviewer 3: How did you arrive at the values (20,000, 20,000, 20,000)

Reviewer 3: Instead of " Each cell in the array represents a cube of 1 light year per side and contains the number of stars within it.", I would suggest "Each cell represents the number of stars present in a cube of size 1 light year"

> The number of stars is stored as a uint8, resulting in a total dataset size of 7.3 TB. However, compression can greatly reduce its size to 6.5 GB since the 3D array is very sparse. Blosc2 can compress the zeroed parts almost entirely.

Reviewer 3: Suggest the rewrite of "The number of stars is stored as a uint8, resulting in a total dataset size of 7.3 TB." as "Ideally, if the total size of the 3-dimensional array, representing the number of stars, stored as a uint8 would be 7.3 TB."

Reviewer 3: How did you arrive at value of "6.5 GB"?

Reviewer 3: This paragraph is the result of the Bloc2 algorithm. Hence should be discussed after the Bloc2 algorithm is explained.

> In addition, we store other data about the stars in a separate table indexed with the position of each star (using PyTables). For demonstration purposes, we store the radial velocity, effective temperature, and G-band magnitude using a float32 for each field. The size of the table is 14.1 GB uncompressed, but it can be compressed to 5.6 GB. Adding another 1.5 GB for the index brings the total size to 7.2 GB. Therefore, the 3D array is 6.5 GB, and the table with the additional information and its index are 7.2 GB, making a total of 13.7 GB. This comfortably fits within the storage capacity of any modern laptop.

Reviewer 3: Again, please provide some detail about the Bloc2 algorithm. Or discuss the results in a section separate from the dataset section.

Reviewer 3: The values of compression do not make sense. The authors should explain here how many values were stored by the algorithm and how was did they arrive at the calculations.

.. figure:: 3d-view-milkyway-inverse.png
   :scale: 25%

   Gaia DR3 dataset as a 3D array (Gaia collaboration). :label:`gaia-3d-dset`

> Figure :ref:`gaia-3d-dset` shows a 3D view of the Milky Way different type of stars. Each point is a star, and the color of each point represents the star's magnitude, with the brightest stars appearing as the reddest points. Although this view provides a unique perspective, the dimensions of the cube are not enough to fully capture the spiral arms of the Milky Way.

Reviewer 3: This should be above the results

Author: Agreed.  Fixed in 48d4341

> One advantage of using a 3D array is the ability to utilize Blosc2 NDim's powerful slicing capabilities for quickly exploring parts of the dataset. For example, we could search for star clusters by extracting small cubes as NumPy arrays, and counting the number of stars in each one. A cube containing an abnormally high number of stars would be a candidate for a cluster. We could also extract a thin 3D slice of the cube and project it as a 2D image, where the pixels colors represent the magnitude of the shown stars. This could be used to generate a cinematic view of a journey over different trajectories in the Milky Way.

Reviewer 3: This should be explained after Blosc2 NDIM explanation.

Blosc2 NDim
-----------

> Blosc2 NDim is a new feature of Blosc2 that allows to create and read n-dimensional datasets in an extremely efficient way thanks to a completely general n-dim 2-level partitioning, allowing to slice and dice arbitrary large (and compressed!) data in a more fine-grained way. Having a second partition provides a better flexibility to fit the different partitions at the different CPU cache levels, making compression even more efficient.

Reviewer 4: ```suggestion
Blosc2 NDim is a new feature of Blosc2 that allows one to create and read n-dimensional datasets in an extremely efficient way thanks to a completely general n-dim 2-level partitioning, letting one slice and dice arbitrary large (and compressed!) data in a more fine-grained way. Having a second partition provides more flexibility to fit different partitions at different CPU cache levels, making compression even more efficient.
```
nits: grammar for better readability

> Blosc2 NDim is a new feature of Blosc2 that allows to create and read n-dimensional datasets in an extremely efficient way thanks to a completely general n-dim 2-level partitioning, allowing to slice and dice arbitrary large (and compressed!) data in a more fine-grained way. Having a second partition provides a better flexibility to fit the different partitions at the different CPU cache levels, making compression even more efficient.

Reviewer 3: "!" should not be used in an article. I suggest "allowing for slicing and dicing arbitrary large and compressed data"

Author: This has gone already

.. figure:: read-partial-slices.png
   :scale: 70%

   Speed comparison for reading partial n-dimensional slices of a 4D dataset. The legends labeled "DIM N" refer to slices taken orthogonally to each dimension. The sizes for the two partitions have been chosen such that the first partition fits comfortably in the L3 cache of the CPU (Intel i9 13900K), and the second partition fits in the L1 cache of the CPU. :cite:`BDT23-blosc2-ndim-intro`. :label:`read-partial-slices`

It is important to note that Blosc2 NDim supports all data types in NumPy. This means that, in addition to the typical data types like signed/unsigned int, single and double-precision floats, bools or strings, it can also store datetimes (including units), and arbitrarily nested heterogeneous types. This allows to create multidimensional tables and more.

> Support for multiple codecs, filters, and other compression features

Reviewer 3: This section does not seem to be relevant to the paper. The author should mention somewhere how this is used for the Milky Way dataset.

Author: That may be true.  However, I still do think there is value is describing the capabilities of the library, even if not everything is used in the data handling in the article.

.. figure:: blosc2-pipeline-v2.png

   The Blosc2 filter pipeline. During compression, the first function applied is the prefilter (if any), followed by the filter pipeline (with a maximum of six filters), and finally, the codec. During decompression, the order is reversed: first the codec, then the filter pipeline, and finally the postfilter (if any). :label:`blosc2-pipeline`

Furthermore, Blosc2 supports user-defined codecs and filters, allowing one to create their own compression algorithms and use them within Blosc2 :cite:`BDT22-blosc2-pipeline`. These user-defined codecs and filters can also be dynamically loaded :cite:`BDT23-dynamic-plugins`, registered globally within Blosc2, and installed via a Python wheel so that they can be used seamlessly from any Blosc2 application (whether in C, Python, or any other language that provides a Blosc2 wrapper).

> Automatic tuning of compression parameters

Reviewer 3: The tuning was not mentioned anywhere before. A line should be mentioned in the abstract about tuning using neural networks.

Author: Fair point.  Fixed in a6bf6340

------------------------------------------

> Finding the right compression parameters for the data is probably the most difficult part of using a compression library. Which combination of code and filter would provide the best compression ratio? Which one would provide the best compression/decompression speed?

Reviewer 4: ```suggestion
Finding the right compression parameters for the data is probably the most difficult part of using a compression library. Which combination of code and filters would provide the best compression ratio? Which one would provide the best compression/decompression speed?
```

Author: Addressed in 64e5ea67

> BTune is an AI tool for Blosc2 that automatically finds the optimal combination of compression parameters to suit the user needs. It uses a neural network that is trained on the most representative datasets to be compressed. This allows it to predict the best compression parameters based on a given balance between compression ratio and compression/decompression speed.

Reviewer 4: ```suggestion
BTune is an AI tool for Blosc2 that automatically finds the optimal combination of compression parameters to suit user needs. It uses a neural network trained on representative datasets to be compressed to predict the best compression parameters based on the given tradeoff between compression ratio and compression/decompression speed.
```

Author: Addressed in 64e5ea67
